File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/B.Комбинации/solution.py
import sys


keyboard: dict[str, str] = {
    '2' : 'abc', '3' : 'def', '4' : 'ghi',
    '5' : 'jkl', '6' : 'mno', '7' : 'pqrs',
    '8' : 'tuv', '9' : 'wxyz'
}


printed_first_time: bool = False


def backtracking(idx: int, buttons: str, typing: list[str]) -> None:
    n: int = len(buttons)
    if idx == n:
        global printed_first_time
        if printed_first_time:
            sys.stdout.write(' ')
        sys.stdout.write(''.join(typing))
        printed_first_time = True
        return
    button: str = buttons[idx]
    for key in keyboard[button]:
        typing.append(key)
        backtracking(idx + 1, buttons, typing)
        typing.pop()


def solution(buttons: str) -> None:
    """
    Time Complexity O(N**K), where N is length of buttons and K is number of keyboards
    Memory Complexity O(N)
    """
    printed_first_time = False
    backtracking(0, buttons, [])


if __name__ == '__main__':
    buttons: str = sys.stdin.readline().rstrip()
    solution(buttons)

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/B.Комбинации/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>


std::unordered_map<char, std::string> keyboard {
    {'2', "abc"}, {'3', "def"}, {'4', "ghi"},
    {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"},
    {'8', "tuv"}, {'9', "wxyz"}
};


bool printedFirstTime = false;


void backtracking(int index, std::string& buttons, std::string& typing) {
    int n = buttons.size();
    if (index == n) {
        if (printedFirstTime) {
            std::cout << " ";
        }
        std::cout << typing;
        printedFirstTime = true;
    }
    char button = buttons[index];
    for (char key : keyboard[button]) {
        typing += key;
        backtracking(index + 1, buttons, typing);
        typing.pop_back();
    }
}


void solution(std::string& buttons) {
    /*
    Time Complexity O(N**K), where N is length of buttons and K is number of keyboards
    Memory Complexity O(N)
    */
    printedFirstTime = false;
    std::string typing = "";
    backtracking(0, buttons, typing);
}


int main() {
    std::string buttons;
    std::cin >> buttons;
    solution(buttons);
}